src/Examine/LuceneEngine/Providers/LuceneIndex.cs

```diff
@@ -43,8 +44,54 @@ public class LuceneIndex : BaseIndexProvider, IDisposable, IIndexStats
             Analyzer analyzer = null,
             IValueSetValidator validator = null,
             IReadOnlyDictionary<string, IFieldValueTypeFactory> indexValueTypesFactory = null)
-            : this(name, luceneDirectory, DefaultQueueCapacity, fieldDefinitions, analyzer, validator, indexValueTypesFactory)
+            : this(name, luceneDirectory, DefaultQueueCapacity, fieldDefinitions, analyzer, validator,
+                indexValueTypesFactory)
+        {
+            if (luceneDirectory is ExamineDirectory dir)
+            {
+                isReadonly = dir.IsReadOnly;
+                ExamineDir = dir;
+                dir.HandleOutOfSyncEvent += DirOnHandleOutOfSyncEvent;
+            }
+        }
+
+        public ExamineDirectory ExamineDir { get; set; }
+
+        private void DirOnHandleOutOfSyncEvent(object sender, EventArgs e)
         {
+            lock (ExamineDir.RebuildLock)
```

Comment:
What else is using this lock? This lock won't work with the _writerLocker, what happens if somehow this event is fired while the writer is being created or indexing is occuring. You need to lock on _writerLocker too since you are modifying the _writer.

Reply:
Maybe it's better to have a private method to "Reset"? that is more descriptive of what it is doing and then have this handler call that method.